### PR TITLE
refactor(compiler): cleanup and preparation for integration

### PR DESCRIPTION
**Pull request by tbosch**
- Rename `DirectiveMetadata` into `CompileDirectiveMetadata`, merge
  with `NormalizedDirectiveMetadata` and remove `ChangeDetectionMetadata`
- Store change detector factories not as array but
  directly at the `CompiledTemplate` or the embedded template
  to make instantiation easier later on
- Already analyze variable values and map them
  to `Directive.exportAs`
- Keep the directive sort order as specified in the
  `@View()` annotation
- Allow to clear the runtime cache in `StyleCompiler`
  and `TemplateCompiler`
- Ignore `script` elements to match the semantics of the
  current compiler
- Make all components dynamically loadable and remove
  the previously introduced property `@Component#dynamicLoadable`
  for now until we find a better option to configure this
- Don’t allow to specify bindings in `@View#directives` and `@View#pipes` as this was never supported by the transformer (see below for the breaking change)

BREAKING CHANGE:
- don't support DI bindings in `@View#directives` and `@View@pipes` any more in preparation of integrating the new compiler. Use `@Directive#bindings` to reexport directives under a different token instead.

Part of #3605

### CODE REVIEW

File: modules/angular2/test/compiler/template_parser_spec.ts [hunk 1]
```diff
@@ -512,6 +476,83 @@ export function main() {
...
+
+        it('should parse variables via var-... and not report them as attributes', () => {
+          expect(humanizeTemplateAsts(parse('<div var-a>', [])))
+              .toEqual([
+                [ElementAst, 'div', 'TestComp > div:nth-child(0)'],
+                [VariableAst, 'a', '', 'TestComp > div:nth-child(0)[var-a=]']
+              ]);
+        });
+
+        it('should camel case variables', () => {
```
Comment (pkozlowski-opensource) [inline modules/angular2/test/compiler/template_parser_spec.ts]: @tbosch now that we don't run the compiler on the DOM parsed by a browser, what are you thoughts on relaxing case-insensitivity rules imposed by a browser. In this particular example, do we want to let people write `var-someA` and be it equivalent to `var-some-a`? 

Same applies to property names.

File: modules/angular2/src/compiler/template_parser.ts [hunk 2]
```diff
@@ -154,6 +163,14 @@ class TemplateParseVisitor implements HtmlAstVisitor {
 
   visitElement(element: HtmlElementAst, component: Component): any {
     var nodeName = element.name;
+    var preparsedElement = preparseElement(element);
+    if (preparsedElement.type === PreparsedElementType.SCRIPT ||
+        preparsedElement.type === PreparsedElementType.STYLE ||
+        preparsedElement.type === PreparsedElementType.STYLESHEET ||
+        preparsedElement.type === PreparsedElementType.NON_BINDABLE) {
+      return null;
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Add comment why we skip these

Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Done

File: modules/angular2/test/compiler/template_parser_spec.ts (hunk 1 above)
Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: Could you create an issue explaining the details?

Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: @pkozlowski-opensource Well, the compiler can still be run in the browser, in which case we will use the regular DOM compiler...

File: modules/angular2/src/compiler/template_parser.ts [hunk 3]
```diff
@@ -58,10 +64,10 @@ var BIND_NAME_REGEXP =
 
 const NG_CONTENT_ELEMENT = 'ng-content';
```
Comment (tbosch) [inline modules/angular2/src/compiler/template_parser.ts]: Remove these!

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 4]
```diff
@@ -11,6 +12,21 @@ export enum LifecycleHooks {
   AfterViewChecked
 }
 
+export var lifecycleHookMap: Map<number, LifecycleHooks> = MapWrapper.createFromPairs([
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/interfaces.ts]: Use a plain list, also for the other enums.

File: modules/angular2/src/core/compiler/interfaces.ts [hunk 5]
```diff
@@ -11,6 +12,21 @@ export enum LifecycleHooks {
...
+  [1, LifecycleHooks.OnDestroy],
+  [2, LifecycleHooks.DoCheck],
+  [3, LifecycleHooks.OnChanges],
+  [4, LifecycleHooks.AfterContentInit],
+  [5, LifecycleHooks.AfterContentChecked],
+  [6, LifecycleHooks.AfterViewInit],
+  [7, LifecycleHooks.AfterViewChecked]
+]);
+
+export function lifecycleHookFromJson(value: number): LifecycleHooks {
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/interfaces.ts]: remove this method, also for the other enums...

File: modules/angular2/test/compiler/template_parser_spec.ts (hunk 1 above)
Comment (vicb) [inline modules/angular2/test/compiler/template_parser_spec.ts]: @pkozlowski-opensource  IMO deviating from the standard would really be confusing here. Also parse5 lowercase attribute names.

@tbosch does that mean that we will have 2 different compilers - the regular and the offline ones ?

File: modules/angular2/src/core/compiler/template_commands.ts [hunk 6]
```diff
@@ -43,13 +43,11 @@ export class CompiledTemplate {
 
 const EMPTY_ARR = CONST_EXPR([]);
 
-export interface TemplateCmd extends RenderTemplateCmd {
-  visit(visitor: CommandVisitor, context: any): any;
-}
+export interface TemplateCmd extends RenderTemplateCmd { visit(visitor: any, context: any): any; }
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/template_commands.ts]: try to use generics again, maybe just have one visitor...

File: modules/angular2/src/core/compiler/template_commands.ts [hunk 7]
```diff
@@ -69,30 +65,29 @@ export function ngContent(ngContentIndex: number): NgContentCmd {
 }
 
 export interface IBeginElementCmd extends TemplateCmd, RenderBeginElementCmd {
-  variableNameAndValues: string[];
-  eventNames: string[];
+  variableNameAndValues: any[];
```
Comment (tbosch) [inline modules/angular2/src/core/compiler/template_commands.ts]: use `<string | number>[]`

Comment (tbosch) [inline modules/angular2/src/core/compiler/template_commands.ts]: `Array<string | number>`

File: modules/angular2/test/compiler/command_compiler_spec.ts [hunk 8]
```diff
@@ -226,8 +255,8 @@ export function main() {
                        BEGIN_COMPONENT,
                        'a',
                        ['a', 'b'],
-                       ['click'],
-                       ['someVar', 'someValue'],
+                       [null, 'click'],
+                       ['someVar', '%implicit', 'someVar', 0],
```
Comment (tbosch) [inline modules/angular2/test/compiler/command_compiler_spec.ts]: component should take precedence...

File: modules/angular2/test/compiler/template_parser_spec.ts (hunk 1 above)
Comment (pkozlowski-opensource) [inline modules/angular2/test/compiler/template_parser_spec.ts]: @tbosch @vicb right, I'm also not keen on deviating from the standard. But:
* people are having problems with snake-case to camelCase convention
* I was under the impression that we are going to use the same compiler in a browser an during the build. In other words I was assuming that we parse strings and not the live DOM, even in the browser.

So, I second @vicb question here: will we have 2 different compilers - the regular and the offline ones ?

**Comment by tbosch** [comment]
Reviewed in person with @vsavkin

**Comment by tbosch** [comment]
Sorry, wrong person. Reviewed in person with @yjbanov

File: modules/angular2/test/compiler/template_parser_spec.ts (hunk 1 above)
Comment (tbosch) [inline modules/angular2/test/compiler/template_parser_spec.ts]: The new compiler has 2 code paths: for runtime and for codegen. These code paths share most of their implementation and also tests! The main difference is that the runtime compiler does not codegen but creates the artifacts directly, i.e. by calling `new` on them.

**Pull request closed without merging**